pkg/ext/delegate.go

```diff
@@ -0,0 +1,337 @@
+package ext
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+type typeChecker struct {
+	runtime.Object
+}
+
+type typeCheckerList struct {
+	runtime.Object
+}
+
+// XXX: Implement DeleteCollection to simplify everything here
+// var _ rest.StandardStorage = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Storage = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Scoper = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.KindProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.GroupVersionKindProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.SingularNameProvider = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+
+var _ rest.Getter = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Lister = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.GracefulDeleter = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Creater = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Updater = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+var _ rest.Watcher = (*delegate[*typeChecker, typeChecker, *typeCheckerList, typeCheckerList])(nil)
+
+// Ptr[U] acts as a type constraint such that
+//
+//	T Ptr[U]
+//
+// means that T is a pointer to U and a runtime.Object.
+type Ptr[U any] interface {
+	*U
+	runtime.Object
+}
+
+// delegate is the bridge between k8s.io/apiserver's [rest.Storage] interface and
+// our own Store interface we want developers to use
+//
+// It is used for non-namespaced objects only.
```

Comment:
With the re-organization of a few things (like type assertions, etc), the delegate is now at the top of the file.

> can you elaborate on why this isn't necessary for namespaced objects?

As discussed, the delegate only supports non-namespaced resources because the `Store` themselves have no concept of namespaces (eg: `Get()` is given a name, but not a namespace). So until we support stores with namespaces, the delegate won't support namespaced resources.

I have updated the comment with the following, let me know if that is less confusing.

```go
  // It currently supports non-namespaced stores only because Store[T, TList] doesn't                                                                                                                                                                                  
  // expose namespaces anywhere. When needed we'll add support to namespaced resources.
```